### ISSUE TITLE
Document why some vars are not setuid safe

### ISSUE THREAD

**Issue by FooBarWidget** [fullstaq-labs/activatecontaineruid#2]

**Comment by mbrodala** [comment]
This should indeed be explained in detail, especially why basically all environment variables are ignored:

```
app-1  | 2026-06-18 06:57:25,949 WARN  [matchhostfsowner::config] Ignoring MHF_HOST_UID env var because we got root privileges via the setuid root bit.
app-1  | 2026-06-18 06:57:25,949 WARN  [matchhostfsowner::config] Ignoring MHF_HOST_GID env var because we got root privileges via the setuid root bit.
app-1  | 2026-06-18 06:57:25,949 WARN  [matchhostfsowner::config] Ignoring MHF_APP_ACCOUNT env var because we got root privileges via the setuid root bit.
app-1  | 2026-06-18 06:57:25,949 WARN  [matchhostfsowner::config] Ignoring MHF_APP_GROUP env var because we got root privileges via the setuid root bit.
app-1  | 2026-06-18 06:57:25,950 ERROR [matchhostfsowner] Error looking up details for OS user account 'app': User not found in user database
```

This basically renders this tool useless if the host UID/GID should be determined at runtime. A static configuration file is not an option in this case.

**Comment by FooBarWidget** [comment]
The general security reasoning goes like this: if it's setuid root, then any unprivileged user can run it with malicious input, potentially resulting in privilege escalation. Therefore I need to be paranoid about trusting input.

A root-owned config file is considered trusted. Environment variables can be forged at will.

Would it be possible for you to dynamically create a config file in the container entrypoint?

**Comment by mbrodala** [comment]
But you can check if your tool runs as `pid=1` in which case it is a trusted execution. Here environment variables can only be set by the operator since no runtime code runs before. Also you already drop the `setuid` bit afterwards so any manipulation attempt afterwards is moot.

Or am I missing something?

**Comment by FooBarWidget** [comment]
I need to think about this longer. My first instinct tells me that pid 1 does not necessarily mean safe/trusted. It could be an unprivileged script that then re-execs itself as matchhostfsowner. This retains pid 1, but matchhostfsowner could then be in a malicious environment.

**Comment by mbrodala** [comment]
And I didn't answer:

> Would it be possible for you to dynamically create a config file in the container entrypoint?

Unfortunately no since the goal is to adjust the container user UID/GID at runtime based on the UID/GID of a specific directory which is mounted from the host. And the container runs as unprivileged user so it cannot write to `/etc`.

**Comment by mbrodala** [comment]
Even more background: when running the container as `root` the UID/GID adjustment based on the `MHF_*` environment variables works.

But then I get other issues like `/proc/self/fd/*` being owned by `root` and thus not writable by the container user and `docker compose exec <service> <command>` running as `root`.

**Comment by FooBarWidget** [comment]
The recommended way to use matchhostfsowner in setuid mode ("mode 1") is to run the container with `--user "<desired UID>:<desired GID>`. Then you don't need to set MHF_HOST_UID/GID. You just need to create a config file at build time to ensure the app UID/GID is known (if it's not "app"). Does this work for you?

**Comment by mbrodala** [comment]
Yes, I read the docs. And as described this is not an option for me:

1. I do not know the host UID/GID at build time
2. I need to determine the target UID/GID at runtime based on the UID/GID of a specific directory
3. The container should run with its default `USER` which is not `root`
4. I need to adjust the UID/GID of this very `USER` at runtime, then run the original entrypoint/command

**Comment by FooBarWidget** [comment]
> I do not know the host UID/GID at build time

But the solution I described (`docker run --user "<desired UID>:<desired GID>"`) allows you to set the target UID/GID at runtime, not build time. Is there some misunderstanding here?

**Comment by mbrodala** [comment]
Yes, but I'd like to avoid that. Developers should just `docker compose up` without worrying about anything like this. As mentioned the desired UID/GID can be determined by inspecting (`stat`) the mounted volume since it is owned by the developer.

**Comment by mbrodala** [comment]
OK, just tried with this `docker-compose.override.yml`:

```yaml
services:
  app:
    user: ${UID:-www-data}:${GID:-www-data}
```
plus UID / GID in `.env`

And it just works ™️ . I guess it's better to bite this bullet than continue wasting more time for automatic.

Thanks for the hints!
